### DISCUSSION TITLE
[Proposal]: simplify testing multiple equality with in comparison

### DISCUSSION THREAD

**Discussion by softlion** [dotnet/csharplang#4065] [Language Ideas] [answered]
# In comparison

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

This code:
```csharp
IsRotating = args.Status == GestureStatus.Started ||  args.Status == GestureStatus.Running
```

could be simplified to:
```csharp
IsRotating = args.Status in [GestureStatus.Started, GestureStatus.Running]
```

like in the SQL language.

## Motivation
[motivation]: #motivation

Cleaner code with all the benefits by stopping repeating the equality comparison and the first argument to this comparison.

## Detailed design
[design]: #detailed-design

This could be expanded by the compiler into the original syntax.

## Drawbacks
[drawbacks]: #drawbacks

"in" is a new keyword there. But it should not impact the existing syntaxes.

## Alternatives
[alternatives]: #alternatives

None ?

## Unresolved questions
[unresolved]: #unresolved-questions

In this the correct keyword to use ?

**Comment by svick** [comment, marked as answer]
C# 9.0 includes `or` patterns, which allow you to write [code like the following](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8A+AAgEwAEAwgLABQA3jaY6YQCykCyAFAOJywCuCOAGUYEGPyilY4yQEoGTetSarSaCAlKYoAJQD2YmJgB2Ac1IBeRWrUyJUnaV4Cho2VAB07hPAAmpPpaLhJuRpKeuvwmJqZmANw2pAC+NKnUNHAm/GDOfKEi4VA0yqo+/kikUTFxaUA===):

```c#
IsRotating = args.Status is GestureStatus.Started or GestureStatus.Running;
```

Is that sufficient for your needs?

**Reply by softlion** [reply]
This syntax is not good. Because « as » is used to compare to a type, not to the value in one type. And because « or » is not the normal or operator.

This makes this syntax bad, badly discoverable and badly understandable, while adding complexity for beginners.

I liked most of your changes.
But for c#9 you are all wrong there guys.

**Reply by CyrusNajmabadi** [reply]
that same argument applies to your syntax. `in` already means something else in c#. As does [] in an expression context. 

Los of c# syntax is reused in different places to mean different things. That's ok and is a standard part of the language. You're doing that yourself with your proposal. :-)

**Reply by theunrepentantgeek** [reply]
@softlion the **is** and **as** operators were extended in C# 8 into pattern operators, and in C# 9 they can test for patterns involving values as well. They are no longer merely doing type comparisons. The syntax shown by @svick is _already available_ in preview, and is on track to be generally released within weeks.
